go/libkb/secret_store_file.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package libkb
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+var ErrSecretForUserNotFound = NotFoundError{Msg: "No secret found for user"}
+
+type SecretStoreFile struct {
+	dir string
+}
+
+var _ SecretStoreAll = (*SecretStoreFile)(nil)
+
+func NewSecretStoreFile(dir string) *SecretStoreFile {
+	return &SecretStoreFile{dir: dir}
+}
+
+func (s *SecretStoreFile) RetrieveSecret(username NormalizedUsername) ([]byte, error) {
+	secret, err := ioutil.ReadFile(s.userpath(username))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrSecretForUserNotFound
+		}
+
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+func (s *SecretStoreFile) StoreSecret(username NormalizedUsername, secret []byte) error {
+	final := s.userpath(username)
+	tmp := final + ".tmp"
+	if err := ioutil.WriteFile(tmp, secret, 0600); err != nil {
+		return err
+	}
+	if err := os.Rename(tmp, final); err != nil {
+		return err
+	}
+	return os.Chmod(final, 0600)
+}
+
+func (s *SecretStoreFile) ClearSecret(username NormalizedUsername) error {
+	return os.Remove(s.userpath(username))
+}
+
+func (s *SecretStoreFile) GetUsersWithStoredSecrets() ([]string, error) {
+	files, err := filepath.Glob(filepath.Join(s.dir, "*.ss"))
+	if err != nil {
+		return nil, err
+	}
+	users := make([]string, len(files))
+	for i, f := range files {
+		users[i] = stripExt(filepath.Base(f))
+	}
+	return users, nil
+}
+
+func (s *SecretStoreFile) userpath(username NormalizedUsername) string {
+	return filepath.Join(s.dir, fmt.Sprintf("%s.ss", username))
```

Comment:
Windows has an unusually large forbidden set, `\/:*?"<>|` I think. We also have to worry about two users with the same name but different capitalization clobbering each other on Windows/OSX. Again something that the server should prevent, but where there might be bugs or security holes around the enforcement.

Then again, is it only username that you're logging in as, which will end up here? If we're sure it'll always be that way, then I guess you can only screw yourself, and it's not so bad.